SimG4CMS/HGCalTestBeam/plugins/HGCalTimingAnalyzer.cc

```diff
@@ -0,0 +1,351 @@
+#include "Geometry/Records/interface/HcalRecNumberingRecord.h"
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "SimDataFormats/CaloHit/interface/PCaloHitContainer.h"
+#include "SimDataFormats/CaloTest/interface/HGCalTestNumbering.h"
+#include "SimDataFormats/GeneratorProducts/interface/HepMCProduct.h"
+#include "SimDataFormats/HcalTestBeam/interface/HcalTestBeamNumbering.h"
+#include "SimDataFormats/Track/interface/SimTrackContainer.h"
+#include "SimDataFormats/Vertex/interface/SimVertexContainer.h"
+#include "SimG4CMS/HGCalTestBeam/interface/AHCalDetId.h"
+
+// Root objects
+#include "TROOT.h"
+#include "TSystem.h"
+#include "TTree.h"
+
+//#define EDM_ML_DEBUG
+
+class HGCalTimingAnalyzer : public edm::one::EDAnalyzer<edm::one::WatchRuns,edm::one::SharedResources> {
+
+public:
+  explicit HGCalTimingAnalyzer(edm::ParameterSet const&);
+  ~HGCalTimingAnalyzer();
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  virtual void beginJob() override ;
+  virtual void beginRun(edm::Run const&, edm::EventSetup const&) override;
+  virtual void endRun(edm::Run const&, edm::EventSetup const&) override {}
+  virtual void analyze(edm::Event const&, edm::EventSetup const&) override;
+  void analyzeSimHits(int type, std::vector<PCaloHit> const& hits);
+  void analyzeSimTracks(edm::Handle<edm::SimTrackContainer> const& SimTk, 
+			edm::Handle<edm::SimVertexContainer> const& SimVtx);
+
+  edm::Service<TFileService>                fs_;
+  const HGCalDDDConstants                  *hgcons_;
+  bool                                      doTree_, groupHits_;
+  std::string                               detectorEE_, detectorBeam_;
+  double                                    timeUnit_;
+  std::vector<int>                          idBeams_;
+  edm::EDGetTokenT<edm::PCaloHitContainer>  tok_hitsEE_, tok_hitsBeam_;
+  edm::EDGetTokenT<edm::SimTrackContainer>  tok_simTk_;
+  edm::EDGetTokenT<edm::SimVertexContainer> tok_simVtx_;
+  edm::EDGetTokenT<edm::HepMCProduct>       tok_hepMC_;
+  TTree                                    *tree_;
+  std::vector<uint32_t>                     simHitCellIdEE_, simHitCellIdBeam_;
+  std::vector<float>                        simHitCellEnEE_, simHitCellEnBeam_;
+  std::vector<float>                        simHitCellTmEE_, simHitCellTmBeam_;
+  double                                    xBeam_, yBeam_, zBeam_, pBeam_;
+};
+
+HGCalTimingAnalyzer::HGCalTimingAnalyzer(const edm::ParameterSet& iConfig) {
+
+  usesResource("TFileService");
+
+  //now do whatever initialization is needed
+  detectorEE_  = iConfig.getParameter<std::string>("DetectorEE");
+  detectorBeam_= iConfig.getParameter<std::string>("DetectorBeam");
+  // Group hits (if groupHits_ = true) if hits come within timeUnit_
+  groupHits_   = iConfig.getParameter<bool>("GroupHits");
+  timeUnit_    = iConfig.getParameter<double>("TimeUnit");
+  // Only look into the beam counters with ID's as in idBeams_
```

Comment:
Yes - it will be. One needs to study timing from one special beam counter for which the SD class is to be revised. It will be added shortly (need quite a bit of work). Meanwhile people will start analyzing timing information from the silicon detector